### ISSUE TITLE
Source uploads

### ISSUE THREAD

**Issue by yhahn** [mapbox/mapbox-studio-classic#472]
Per @camilleanne's notes:
- After export finishes, offer user **Download** or **Upload to Mapbox** options
- Upload job should ideally fit into the same API as export jobs (https://github.com/mapbox/tm2/blob/master/lib/tm.js#L302)
- Ideally we can reuse large portions of code for representing progress, displaying it, etc.
### TODOs
- [x] @yhahn Look into switching TM2 to tilelive streams https://github.com/mapbox/tilelive.js/pull/72
- [x] Normalize how progress reporting is done between https://github.com/mapbox/mapbox-upload/blob/master/index.js and https://github.com/mapbox/tilelive.js/pull/72
- [x] Turn https://github.com/mapbox/tm2/blob/master/lib/tm.js#L302 into an API for any long-running jobs with standardized progress reporting. Could possibly be used for huge static image stitching jobs in the future

**Comment by yhahn** [comment]
## Normalizequest

Goals: return a standard stream to track progress. Drop custom progress stuffs and be compatible with [progress-stream](https://github.com/freeall/progress-stream).

### mapbox-upload refactor

https://github.com/mapbox/mapbox-upload

Drop the custom task eventemitter that's returned and just pass back the main file readstream. Let the start/end actions (getting creds, putting map doc) be secondary to this stream.
@camilleanne: 
- [x] upload(opts, callback) returns _not_ a custom task object but the ReadStream from fs.createReadStream() of the passed in filepath.
- ~~stat the file and emit `length` event to let API users know what the length of the stream will be. Add length to `.length` of stream.~~
- ~~handle readstream.unpipe() elegantly -- e.g. gives API users a way to cancel uploads~~
- ~~revamp tests~~
  (see below)

Flow will now look like

```
upload(opts)
- Synchronous. Returns the readstream from fs.createReadStream
- Readstream is created but not used for piping/reading anything *yet*
- fs.stat the file next, emit length event and add .length key
- getcreds() gets upload credentials
- start piping file to request (the putfile part)
- finish with putmap, emit a 'complete' event from the *file readstream* at the end
```

### tilelive.js

Stream refactor puts us in good place to do this.
- [x] Add `length` event and .length key addition to each readstream
- [x] Update tests
- [x] Use progress-stream in copy bin

### tm2

**pre-refactoring**
- [x] Switch to progress-stream branch of tilelive, add progress-stream as a dep
- [x] Use progress-stream to report progress instead of custom setInterval here https://github.com/mapbox/tm2/blob/direct-upload/lib/tm.js#L384-L391
- [x] Update HTTP endpoints, UI to work with progress reporter

**"taskify" api**

https://github.com/mapbox/tm2/issues/492

**mapbox-upload + tasks**
- [x] When starting an upload, call task.get() to make sure there are no other active tasks
- [x] If no other task is active, start the upload and call task.set() on a new task for the upload
- [x] Come up with some way of persisting completed upload state (e.g. `mapid` field + ... ? timestamp? hash of conf at upload time?) and if conditions are met return Done object with destination URL
- [x] Refactor export template to work with both types of tasks (export/upload)

@camilleanne let me know if you want to go over this -- let's carve this up maybe between us and then coordinate back in tm2 land?

**Comment by yhahn** [comment]
@camilleanne https://github.com/mapbox/tm2/pull/492 is merged into `direct-upload` now. After doing that refactor I think i'd recommend adding `progress-stream` to `mapbox-upload` and having _that_ be the readstream returned may be the most promising approach. I think it could replace the task object and just hang out until it gets piped to at upload time.

Similar to the previous approach, things just change slightly:
- [x] upload(opts) returns the progress stream rather than task. Becomes synchronous.
- [x] stat the file and call `progress.setLength` rather than emitting the length event.
- [x] ~~handle readstream.unpipe() elegantly -- e.g. gives API users a way to cancel uploads~~ I think you can just test this to make sure calling unpipe on the progress-stream works. It worked in the context of uploads.
- [x] revamp tests (whatever the above requires)

**Comment by camilleanne** [comment]
@yhahn I'm curious about calling `unpipe` on progress-streams, because it's a writable stream, so it doesn't have that method, right? It's an available event. http://nodejs.org/api/stream.html#stream_event_unpipe

**Comment by yhahn** [comment]
Yeah so the progress stream is a duplex stream,

```
 A => progress-stream => B
```

i.e. it's a writable stream to A and a readable stream to B.

I'm finding that in the case of the export, calling unpipe() on progress-stream stops it from writing to `B` (the MBTiles file). It was an open question whether the progress-stream would then know to stop reading from A. In my tests this is the case -- progress-stream sucks a little more data after unpipe() is called from A but then stops once there is nowhere for the buffered data to go.

Hoping this is also the case when `file => progress-stream => request` : )

**Comment by camilleanne** [comment]
https://github.com/mapbox/mapbox-upload/blob/source-upload/index.js#L111-L112

alright, it seems that the unintended consequence of refactoring for request changed `A -> prog -> B`, to `((file -> prog) -> multipart/form-data) -> request`,  `prog.unpipe()` doesn't seem to stop anything but stop progress-stream from reporting. `prog.end()` causes a "write after end" error.

Pausing the request seems to do the trick, but it's not as elegant.

**Comment by camilleanne** [comment]
Mapbox-Upload  + Tasks have been integrated.

Now, if a source is uploaded to mapbox.com, a mapid is recorded in data.yml. We can check for a corresponding mapid on mapbox.com. If none exists (because the map has been deleted), the source can be uploaded again. If a change has registered in the source, then the source can be uploaded to update that map.

Currently this system has a bunch of holes in the UI and in the backend.
Options shouldn't be:
`[upload new, update upload, already uploaded]`, but probably, `[upload new, update upload, upload with new id]` and maybe `update specific mapid`

Some possible scenarios:
- New source, new upload, new mapid (_supported_)
- ~~New source, new upload, old mapid (e.g. update)~~
- Reexported source, prev upload, old mapid (e.g. update) (_supported_)
- ~~Reexported source, prev upload, new mapid~~
- Old source, new upload, new mapid
- Old source, prev upload deleted/failed, old mapid (_supported in backend, not in ui_)
- ~~Old source, prev upload deleted/failed, new mapid~~

('prev upload' will have an already associated mapid in data.yml and potentially on mapbox.com)

**Comment by yhahn** [comment]
@camilleanne per voice, next up with the goal of getting all this wonderful work into `master`:
- [x] testing export/upload UI in TM2
- [x] mapbox-upload release
- [x] merge into master

**Comment by yhahn** [comment]
That was fun : ) Will open a new ticket once serialized streams is in place upstream in tilelive.

**Comment by ian29** [comment]
:O \o/
